### PR TITLE
storage: implement intent resolution using LockTableIterator

### PR DESCRIPTION
**Pull request by nvb**
Informs #109648.

This commit implements intent resolution (point and ranged) using a `LockTableIterator`, configured to return all locks for the transaction being resolved and no locks from other transactions. This is the first step towards releasing replicated locks during intent resolution.

While switching to a LockTableIterator, the commit is also able to remove separatedIntentAndVersionIter, iterForKeyVersions and mvccGetIntent, which were all used to avoid handing an MVCCMetadata directly to mvccResolveWriteIntent. Instead of continuing to treat intents as interleaved, we switch to handling intents entirely separately from their provisional value during intent resolution, which avoids jumping through these hoops and makes the code simpler.

The change to `TestMVCCResolveTxnRangeResumeWithManyVersions` is immaterial and has to do with the transaction ID filter being applied before the key limit (inside LockTableIterator), instead of after. The new behavior is actually better.

----

One concern I have about this change is that it removes the call to `SeekIntentGE` in `MVCCResolveWriteIntent`, which was added in d1c91e0e to guard against the case where many pebble tombstones from prior intents from different txns on a key surround the intent being resolved. Conceptually, we'd like to push optimizations that avoid scanning over these tombstones into the `LockTableIterator` like we plan to do for skipping over non-conflicting locks. Doing so would benefit all lock strengths. It would also benefit the case where an intent is not found and the seek hits tombstones from prior intents on later versions.

However, it's not clear how to do this with the current Pebble API. Pebble exposes a `SeekGEWithLimit` method, but this "limit" value is expressed as a key and not as a number of steps. How would we construct a limit key to bound the number of tombstones a seek observes before seeking directly to a specific (txn_id, lock_strength) version?

One option would be to seek to specific versions in the `LockTableIterator` when advancing the iterator in cases where the iterator is configured to match a specific txn ID. For example, performing the following translations:
```
SeekGE({Key: k}) -> SeekGE({Key: k, Strength: Intent, TxnID: <txn_id>})
Next()           -> SeekGE({Key: k, Strength: Exclusive, TxnID: <txn_id>})
Next()           -> SeekGE({Key: k, Strength: Shared, TxnID: <txn_id>})
```
Of course, this gets more complicated when some of these locks are not found and the iterator advances past them while seeking. In such cases, we're back to paying the cost of scanning over the tombstones.

If we knew which lock strengths we had acquired on a key, we could avoid some of this cost, but that would require API changes and client buy-in to track lock spans on a per-strength basis.

I'll capture the impact of this change on the following benchmarks and evaluate:
* BenchmarkIntentResolution
* BenchmarkIntentRangeResolution
* BenchmarkIntentScan

Release note: None

### CODE REVIEW

**Review by sumeerbhola — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110324)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`-- commits` line 146 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/110324#-Ne3p-KZ63SeKthKkC6e):*
Is the problem that the LockUpdate does not tell us what locks were held by this txn on the span?
What if we tracked spans for each lock strength in the txn record?

**Review by sumeerbhola — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110324)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`-- commits` line 146 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/110324#-Ne3p-KZ63SeKthKkC6e):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Is the problem that the LockUpdate does not tell us what locks were held by this txn on the span?
What if we tracked spans for each lock strength in the txn record?
</blockquote></details>

I can't remember whether this optimization was prompted by only microbenchmarks. The second bullet in the commit comment is not optimized even before separated intents. Perhaps I was being extra sensitive to intent resolution performance at that time, given intent resolution latency is important for lowering contention.

Another possibility would be to do a series of multiple SeekGEWithLimit calls for each of the strengths. We don't fully optimize this monotonic seek case in `pebble.Iterator` but it is not hard to change that to use next calls under the hood. We would still need to produce a limit key, which is a bit problematic since it needs to be a valid `EngineKey` for `cmp` to work and needs to be exclusive. If we can efficiently construct the immediately smaller txnid, then we can produce a limit key.

**Comment by nvanbenschoten** [comment]
> Is the problem that the LockUpdate does not tell us what locks were held by this txn on the span?
> What if we tracked spans for each lock strength in the txn record?

Right, we don't currently track which locks with which strengths are held on which keys, only that there are locks with _some_ strength held on certain keys. Tracking the strength of locks in the coordinator and then storing locks with their strength in the txn record would let us perform more precise resolution.

> I can't remember whether this optimization was prompted by only microbenchmarks.

From [what I was able to dig up](https://github.com/cockroachdb/cockroach/pull/60622#issuecomment-780165629), it sounds like one of the SQL logic tests that performed ranged intent resolution became significantly slower with separated intents. I don't see any mention of single-key resolution being so slow that we had to optimize it. It makes me wonder whether the first bullet in the commit message was an opportunistic improvement.

I'm running the benchmarks introduced in that commit over this PR to get a better sense for the impact.

> Another possibility would be to do a series of multiple SeekGEWithLimit calls for each of the strengths. We don't fully optimize this monotonic seek case in pebble.Iterator but it is not hard to change that to use next calls under the hood. We would still need to produce a limit key, which is a bit problematic since it needs to be a valid EngineKey for cmp to work and needs to be exclusive. If we can efficiently construct the immediately smaller txnid, then we can produce a limit key.

I've been considering this. I would improve the worst-case performance of single-key resolution in exchange for a more expensive common case (3 seeks instead of 1). I wonder if that's worth it. We already scan over all locks on a key (including tombstones) when writing these intents, so the degenerate quadratic case where a single key is written and resolved in fast succession is quadratic no matter what we do here.

**Review by arulajmani — APPROVED** [review]
:lgtm: 

I'm assuming we want to bottom out on the tombstones discussion before we merge -- I don't have too much to add, but I'll follow along.

> Reviewed 9 of 9 files at r1, 1 of 1 files at r2, 5 of 5 files at r3, 2 of 2 files at r4, 7 of 7 files at r5, 1 of 5 files at r6, 28 of 28 files at r8, 25 of 25 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110324#-:-NecmIFsCSD1wIaUxq4h:bpwdjgb)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`-- commits` line 52 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/110324#-NecoPJa0dP68yxW5zmg):*
>   found and the iterator advances past them while seeking. In such cases,
  we're back to paying the cost of scanning over the tombstones.

I don't fully follow this bit -- could you ELI5 for me?
___
*[`pkg/storage/mvcc.go` line 4574 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/110324#-Nech2aj0sTo7jj_fo7O:-Nech2aj0sTo7jj_fo7P:b-6zromj) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e9651a7dfd6f1fd6acd752e42fb319c802abd/pkg/storage/mvcc.go#L4574)):*
> ```Go
> 
> // mvccResolveWriteIntent is the core logic for resolving an intent. The
> // function accepts the instructions for how to resolve the intent (encoded in
> ```

nit: s/the instructions/instructions?
___
*[`pkg/storage/mvcc.go` line 4578 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/110324#-Necl0M20JXOJc7Ngwb7:-Necl0M20JXOJc7Ngwb8:bd3rxsj) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e9651a7dfd6f1fd6acd752e42fb319c802abd/pkg/storage/mvcc.go#L4578)):*
> ```Go
> // Returns whether the provided intent was resolved (true) or whether the
> // resolution was a no-op (false).
> // REQUIRES: iter surfaces range keys via IterKeyTypePointsAndRanges.
> ```

nit: while we're here, could you add a new line above REQUIRES please?
___
*[`pkg/storage/mvcc.go` line 5129 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/110324#-NecjG5A7TtCyVOb7nDl:-NecjG5B4fM_tqrEmuE4:bi05250) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e9651a7dfd6f1fd6acd752e42fb319c802abd/pkg/storage/mvcc.go#L5129)):*
> ```Go
> 			// We could also compute a tighter nextKey here if we wanted to.
> 			// TODO(nvanbenschoten): this resumeSpan won't be correct if there
> 			// are multiple locks on the same key. What if only of the locks for
> ```

nit: "only one"

Separately, could we include the strength when constructing `lastResolvedKey` as well?
___
*[`pkg/storage/mvcc.go` line 5134 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/110324#-NeckNjyDZT-o-KK4d9Y:-NeckNjyDZT-o-KK4d9Z:b-6iqs3v) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e9651a7dfd6f1fd6acd752e42fb319c802abd/pkg/storage/mvcc.go#L5134)):*
> ```Go
> 			return numKeys, numBytes, &roachpb.Span{Key: lastResolvedKey.Next(), EndKey: intentEndKey}, resumeReason, nil
> 		}
> 		ltEngineKey, err := ltIter.EngineKey()
> ```

[for me] Why can't we use `LockTableKeyVersion` here, like we do in `MVCCResolveWriteIntent`? Does it have something to do with iterator creation (prefix vs. not)? 
___
*[`pkg/storage/mvcc.go` line 5156 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/110324#-NecjRnF7Q5oGbRKfL4c:-NecjRnF7Q5oGbRKfL4d:bvwrhep) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e9651a7dfd6f1fd6acd752e42fb319c802abd/pkg/storage/mvcc.go#L5156)):*
> ```Go
> 		beforeBytes := rw.BufferedSize()
> 		var ok bool
> 		if ltKey.Strength == lock.Intent {
> ```

Is it worth pulling out a `resolveLock` function? Feel free to defer this until we implement lock resolution for other lock strengths.

**Review by sumeerbhola — APPROVED** [review]
:lgtm: 

> Reviewed 4 of 28 files at r8, 10 of 25 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110324#-:-Nee7aKz65MxCL3xT98J:bg7igkv)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`-- commits` line 52 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/110324#-NecoPJa0dP68yxW5zmg):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

>   found and the iterator advances past them while seeking. In such cases,
  we're back to paying the cost of scanning over the tombstones.

I don't fully follow this bit -- could you ELI5 for me?
</blockquote></details>

if the exact key we are looking for is not there, the internal Pebble DELs that have not been compacted away have to be iterated over. See the first bullet in the commit message for https://github.com/cockroachdb/cockroach/commit/d1c91e0e8180eec95c9365a101a1912649bb0dab -- that optimization works when the key is actually there.
___
*[`pkg/storage/mvcc.go` line 5129 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/110324#-NecjG5A7TtCyVOb7nDl:-Nee6WnmB1ZUhsakixMq:b-5ydufu) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e9651a7dfd6f1fd6acd752e42fb319c802abd/pkg/storage/mvcc.go#L5129)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: "only one"

Separately, could we include the strength when constructing `lastResolvedKey` as well?
</blockquote></details>

The resume span and `LockUpdate` is already set up to use `roachpb.Key` so that would be a bigger change. And there is a maximum of 3 locks that the txn could have held on a `roachpb.Key` so fully processing a key is not going to exceed the limit significantly.
___
*[`pkg/storage/mvcc.go` line 5134 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/110324#-NeckNjyDZT-o-KK4d9Y:-Nee62Ox15BZJOLsek0L:b-n35nu1) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e9651a7dfd6f1fd6acd752e42fb319c802abd/pkg/storage/mvcc.go#L5134)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

[for me] Why can't we use `LockTableKeyVersion` here, like we do in `MVCCResolveWriteIntent`? Does it have something to do with iterator creation (prefix vs. not)?
</blockquote></details>

We also need `ltKey.Key` below which we don't know up front since we are resolving a whole range.

**Review by jbowens — COMMENTED** [review]
> Pebble exposes a `SeekGEWithLimit` method, but this "limit" value is expressed as a key and not as a number of steps.

The problem that `-WithLimit` solves can't be expressed as a number of steps, right? My understanding is that it specifically helps in the case where the sought key doesn't exist, so the seek may need to next through an unbounded number of tombstones to arrive at the next live key. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110324#-:-NesfHJB715oUZDhWqLI:bnvgt3i)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @arulajmani, @nvanbenschoten, and @sumeerbhola)

**Review by nvb — COMMENTED** [review]
The related problem that's relevant here is that `NextWithLimit` doesn't have a way to bound the amount of work performed before bailing. That would be useful for the kind of step-a-few-times-then-seek heuristics we have in a few places (including #110754), which balance the best-case cost of Next with the worst-case cost of Seek. If these heuristics are built above the pebble iterator, they don't have a way to limit the number of tombstones they encounter.

Have we considered pushing such a heuristic into Pebble? Something like a `iter.SeekAfterNextWithLimit(key, limit)`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110324#-:-NesgDLI8Zvla_-m6lVb:b-8aedan)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @arulajmani and @sumeerbhola)

___
*[`pkg/storage/mvcc.go` line 4574 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/110324#-Nech2aj0sTo7jj_fo7O:-NesT54u3Wiu6cv9_pwf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e9651a7dfd6f1fd6acd752e42fb319c802abd/pkg/storage/mvcc.go#L4574)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: s/the instructions/instructions?
</blockquote></details>

Done.
___
*[`pkg/storage/mvcc.go` line 4578 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/110324#-Necl0M20JXOJc7Ngwb7:-NesTjxTCXma_lUGPOga:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e9651a7dfd6f1fd6acd752e42fb319c802abd/pkg/storage/mvcc.go#L4578)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: while we're here, could you add a new line above REQUIRES please?
</blockquote></details>

Done.
___
*[`pkg/storage/mvcc.go` line 5129 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/110324#-NecjG5A7TtCyVOb7nDl:-NesUASp2AOutMJE_bmg:b-t5nqea) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e9651a7dfd6f1fd6acd752e42fb319c802abd/pkg/storage/mvcc.go#L5129)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

The resume span and `LockUpdate` is already set up to use `roachpb.Key` so that would be a bigger change. And there is a maximum of 3 locks that the txn could have held on a `roachpb.Key` so fully processing a key is not going to exceed the limit significantly.
</blockquote></details>

Yeah, I think it's cleanest to just keep this key-oriented. But it's a TODO I need to address in https://github.com/cockroachdb/cockroach/pull/110480 before merging that change.
___
*[`pkg/storage/mvcc.go` line 5156 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/110324#-NecjRnF7Q5oGbRKfL4c:-NesUOJC0DtoZ6z-YJp1:b-9ogn1l) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e9651a7dfd6f1fd6acd752e42fb319c802abd/pkg/storage/mvcc.go#L5156)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Is it worth pulling out a `resolveLock` function? Feel free to defer this until we implement lock resolution for other lock strengths.
</blockquote></details>

Yep, see https://github.com/cockroachdb/cockroach/pull/110480 :smiley:

**Review by jbowens — COMMENTED** [review]
> Have we considered pushing such a heuristic into Pebble? Something like a `iter.SeekAfterNextWithLimit(key, limit)`.

`SeekGEWithLimit` will transparently try to next internally for repeated seeks with monotonically increasing keys, but it has limitations. See:
https://github.com/cockroachdb/pebble/blob/bb00504445221cbb49864fea1277905a5ee518dc/iterator.go#L1207-L1249 https://github.com/cockroachdb/pebble/blob/bb00504445221cbb49864fea1277905a5ee518dc/sstable/reader_iter_single_lvl.go#L669-L696

I think this is what motivated @sumeerbhola's three-seek suggestion, but I'm not sure if any of the TrySeekUsingNext limitations apply here.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110324#-:-NetUSPL0KqHL0OhLWDb:b-b7xpzk)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @arulajmani and @sumeerbhola)

**Comment by nvanbenschoten** [comment]
> I'll capture the impact of this change on the following benchmarks and evaluate

Here is the impact that this change has on the benchmarks added in https://github.com/cockroachdb/cockroach/commit/d1c91e0e8180eec95c9365a101a1912649bb0dab. We see that it's non-impactful except for the most degenerate cases, as expected:
```
name                                                  old time/op  new time/op   delta
IntentScan/versions=10/percent-flushed=0-30           2.84µs ± 0%   2.82µs ± 0%    -0.78%  (p=0.000 n=10+9)
IntentScan/versions=10/percent-flushed=50-30          1.61µs ± 0%   1.60µs ± 0%    -1.19%  (p=0.000 n=10+10)
IntentScan/versions=10/percent-flushed=80-30          1.14µs ± 0%   1.13µs ± 0%    -1.21%  (p=0.000 n=10+9)
IntentScan/versions=10/percent-flushed=90-30           943ns ± 0%    934ns ± 0%    -0.95%  (p=0.000 n=10+10)
IntentScan/versions=10/percent-flushed=100-30          895ns ± 0%    888ns ± 0%    -0.75%  (p=0.000 n=10+10)
IntentScan/versions=100/percent-flushed=0-30          33.1µs ± 1%   31.8µs ± 2%    -4.15%  (p=0.000 n=8+10)
IntentScan/versions=100/percent-flushed=50-30         8.56µs ± 4%   8.37µs ± 3%      ~     (p=0.050 n=9+9)
IntentScan/versions=100/percent-flushed=80-30         4.29µs ± 2%   4.28µs ± 2%      ~     (p=0.796 n=10+10)
IntentScan/versions=100/percent-flushed=90-30         2.66µs ± 1%   2.65µs ± 2%      ~     (p=0.056 n=9+10)
IntentScan/versions=100/percent-flushed=100-30         975ns ± 1%    970ns ± 0%    -0.53%  (p=0.001 n=10+9)
IntentScan/versions=200/percent-flushed=0-30          67.5µs ± 1%   65.0µs ± 1%    -3.66%  (p=0.000 n=10+10)
IntentScan/versions=200/percent-flushed=50-30         20.3µs ± 1%   20.1µs ± 8%      ~     (p=0.400 n=9+10)
IntentScan/versions=200/percent-flushed=80-30         7.13µs ± 6%   6.87µs ± 2%    -3.65%  (p=0.003 n=10+10)
IntentScan/versions=200/percent-flushed=90-30         4.26µs ± 3%   4.24µs ± 2%      ~     (p=0.393 n=10+10)
IntentScan/versions=200/percent-flushed=100-30        1.04µs ± 1%   1.04µs ± 1%      ~     (p=0.135 n=10+10)
IntentScan/versions=400/percent-flushed=0-30          33.6µs ± 3%   33.0µs ± 2%    -1.89%  (p=0.001 n=10+9)
IntentScan/versions=400/percent-flushed=50-30         43.9µs ± 1%   43.2µs ± 1%    -1.66%  (p=0.002 n=10+10)
IntentScan/versions=400/percent-flushed=80-30         15.3µs ± 4%   15.4µs ± 7%      ~     (p=0.684 n=10+10)
IntentScan/versions=400/percent-flushed=90-30         7.14µs ± 2%   7.09µs ± 4%      ~     (p=0.161 n=9+9)
IntentScan/versions=400/percent-flushed=100-30        1.77µs ± 2%   1.75µs ± 1%    -0.88%  (p=0.023 n=9+9)
IntentResolution/versions=10/percent-flushed=0-30     3.60µs ± 1%   3.80µs ± 1%    +5.56%  (p=0.000 n=10+10)
IntentResolution/versions=10/percent-flushed=50-30    3.16µs ± 2%   2.55µs ± 1%   -19.27%  (p=0.000 n=10+10)
IntentResolution/versions=10/percent-flushed=80-30    2.61µs ± 1%   2.29µs ± 2%   -12.41%  (p=0.000 n=10+10)
IntentResolution/versions=10/percent-flushed=90-30    2.55µs ± 1%   2.07µs ± 1%   -18.69%  (p=0.000 n=10+9)
IntentResolution/versions=10/percent-flushed=100-30   2.44µs ± 1%   1.98µs ± 1%   -19.04%  (p=0.000 n=10+10)
IntentResolution/versions=100/percent-flushed=0-30    10.5µs ± 5%   18.2µs ± 3%   +73.37%  (p=0.000 n=10+10)
IntentResolution/versions=100/percent-flushed=50-30   4.54µs ± 1%   6.30µs ± 2%   +38.76%  (p=0.000 n=10+10)
IntentResolution/versions=100/percent-flushed=80-30   4.28µs ± 2%   3.84µs ± 1%   -10.26%  (p=0.000 n=10+10)
IntentResolution/versions=100/percent-flushed=90-30   3.90µs ± 1%   3.12µs ± 1%   -19.87%  (p=0.000 n=10+10)
IntentResolution/versions=100/percent-flushed=100-30  2.61µs ± 1%   2.03µs ± 1%   -22.08%  (p=0.000 n=10+9)
IntentResolution/versions=200/percent-flushed=0-30    14.8µs ± 4%   35.9µs ± 2%  +143.54%  (p=0.000 n=9+10)
IntentResolution/versions=200/percent-flushed=50-30   4.86µs ± 2%  10.53µs ± 5%  +116.63%  (p=0.000 n=10+10)
IntentResolution/versions=200/percent-flushed=80-30   4.60µs ± 2%   5.59µs ± 1%   +21.50%  (p=0.000 n=10+9)
IntentResolution/versions=200/percent-flushed=90-30   4.35µs ± 2%   3.90µs ± 1%   -10.32%  (p=0.000 n=10+10)
IntentResolution/versions=200/percent-flushed=100-30  2.65µs ± 1%   2.07µs ± 1%   -21.89%  (p=0.000 n=10+10)
IntentResolution/versions=400/percent-flushed=0-30    5.10µs ± 2%  16.54µs ± 3%  +224.50%  (p=0.000 n=10+10)
IntentResolution/versions=400/percent-flushed=50-30   5.16µs ± 2%  22.19µs ± 2%  +329.77%  (p=0.000 n=10+10)
IntentResolution/versions=400/percent-flushed=80-30   4.92µs ± 0%   8.74µs ± 3%   +77.82%  (p=0.000 n=8+10)
IntentResolution/versions=400/percent-flushed=90-30   4.67µs ± 1%   5.60µs ± 1%   +19.80%  (p=0.000 n=10+10)
IntentResolution/versions=400/percent-flushed=100-30  2.83µs ± 1%   2.10µs ± 1%   -25.91%  (p=0.000 n=10+10)
```
Given that these cases impact the front half of mutations regardless of what we do during intent resolution, I'm ok with the limited regressions on this microbenchmark. Still, to minimize the impact, I re-introduced half of the previous optimization into `MVCCResolveWriteIntent` by initially seeking to the intent version instead of the version prefix. I also left a comment explaining the opportunity to optimize further if we track the strengths that locks on a given key are held with, with a reference back to d1c91e0e. For now, I'll avoid optimizing further until we see this come up in real-world workloads.

TFTRs!

bors r+

**Comment by nvanbenschoten** [comment]
For completeness, here's the impact of this change on `BenchmarkIntentRangeResolution`:
```
name                                                                                               old time/op  new time/op  delta
IntentRangeResolution/versions=10/sparseness=1/other-txn-intents=false/percent-flushed=0-30         256µs ± 1%   263µs ± 2%   +2.95%  (p=0.000 n=10+10)
IntentRangeResolution/versions=10/sparseness=1/other-txn-intents=false/percent-flushed=50-30        117µs ± 1%   124µs ± 1%   +6.11%  (p=0.000 n=10+9)
IntentRangeResolution/versions=10/sparseness=1/other-txn-intents=false/percent-flushed=100-30      68.0µs ± 2%  75.4µs ± 2%  +10.89%  (p=0.000 n=10+10)
IntentRangeResolution/versions=10/sparseness=100/other-txn-intents=false/percent-flushed=0-30       210µs ± 1%   208µs ± 1%   -0.78%  (p=0.000 n=10+9)
IntentRangeResolution/versions=10/sparseness=100/other-txn-intents=false/percent-flushed=50-30     51.6µs ± 1%  51.5µs ± 1%     ~     (p=0.905 n=9+10)
IntentRangeResolution/versions=10/sparseness=100/other-txn-intents=false/percent-flushed=100-30    2.19µs ± 4%  2.28µs ± 2%   +4.09%  (p=0.000 n=10+9)
IntentRangeResolution/versions=10/sparseness=100/other-txn-intents=true/percent-flushed=0-30        237µs ± 1%   203µs ± 0%  -14.13%  (p=0.000 n=9+10)
IntentRangeResolution/versions=10/sparseness=100/other-txn-intents=true/percent-flushed=50-30      81.9µs ± 1%  50.7µs ± 1%  -38.11%  (p=0.000 n=9+9)
IntentRangeResolution/versions=10/sparseness=100/other-txn-intents=true/percent-flushed=100-30     42.1µs ± 1%  12.4µs ± 1%  -70.53%  (p=0.000 n=9+10)
IntentRangeResolution/versions=10/sparseness=1000/other-txn-intents=false/percent-flushed=0-30      209µs ± 1%   206µs ± 1%   -1.72%  (p=0.000 n=10+9)
IntentRangeResolution/versions=10/sparseness=1000/other-txn-intents=false/percent-flushed=50-30    49.7µs ± 1%  49.8µs ± 1%     ~     (p=0.447 n=10+9)
IntentRangeResolution/versions=10/sparseness=1000/other-txn-intents=false/percent-flushed=100-30   1.34µs ± 3%  1.32µs ± 4%     ~     (p=0.051 n=9+10)
IntentRangeResolution/versions=10/sparseness=1000/other-txn-intents=true/percent-flushed=0-30       236µs ± 1%   202µs ± 1%  -14.17%  (p=0.000 n=10+10)
IntentRangeResolution/versions=10/sparseness=1000/other-txn-intents=true/percent-flushed=50-30     81.8µs ± 1%  49.6µs ± 1%  -39.32%  (p=0.000 n=10+10)
IntentRangeResolution/versions=10/sparseness=1000/other-txn-intents=true/percent-flushed=100-30    41.8µs ± 1%  11.5µs ± 5%  -72.35%  (p=0.000 n=9+10)
IntentRangeResolution/versions=100/sparseness=1/other-txn-intents=false/percent-flushed=0-30       2.77ms ± 1%  2.77ms ± 2%     ~     (p=0.604 n=9+10)
IntentRangeResolution/versions=100/sparseness=1/other-txn-intents=false/percent-flushed=50-30       717µs ± 4%   735µs ± 4%   +2.57%  (p=0.035 n=10+10)
IntentRangeResolution/versions=100/sparseness=1/other-txn-intents=false/percent-flushed=100-30     67.3µs ± 1%  74.6µs ± 1%  +10.85%  (p=0.000 n=9+8)
IntentRangeResolution/versions=100/sparseness=100/other-txn-intents=false/percent-flushed=0-30     2.74ms ± 1%  2.75ms ± 2%     ~     (p=0.481 n=10+10)
IntentRangeResolution/versions=100/sparseness=100/other-txn-intents=false/percent-flushed=50-30     651µs ± 6%   632µs ± 3%   -2.99%  (p=0.028 n=10+9)
IntentRangeResolution/versions=100/sparseness=100/other-txn-intents=false/percent-flushed=100-30   2.13µs ± 0%  2.28µs ± 2%   +7.08%  (p=0.000 n=8+10)
IntentRangeResolution/versions=100/sparseness=100/other-txn-intents=true/percent-flushed=0-30      2.74ms ± 1%  2.73ms ± 2%     ~     (p=0.247 n=10+10)
IntentRangeResolution/versions=100/sparseness=100/other-txn-intents=true/percent-flushed=50-30      674µs ± 2%   622µs ± 2%   -7.64%  (p=0.000 n=10+9)
IntentRangeResolution/versions=100/sparseness=100/other-txn-intents=true/percent-flushed=100-30    39.5µs ± 0%  12.4µs ± 1%  -68.70%  (p=0.000 n=8+9)
IntentRangeResolution/versions=100/sparseness=1000/other-txn-intents=false/percent-flushed=0-30    2.73ms ± 2%  2.72ms ± 2%     ~     (p=0.739 n=10+10)
IntentRangeResolution/versions=100/sparseness=1000/other-txn-intents=false/percent-flushed=50-30    639µs ± 6%   628µs ± 6%     ~     (p=0.280 n=10+10)
IntentRangeResolution/versions=100/sparseness=1000/other-txn-intents=false/percent-flushed=100-30  1.35µs ± 3%  1.32µs ± 2%   -2.82%  (p=0.000 n=10+9)
IntentRangeResolution/versions=100/sparseness=1000/other-txn-intents=true/percent-flushed=0-30     2.76ms ± 2%  2.70ms ± 1%   -2.15%  (p=0.000 n=10+9)
IntentRangeResolution/versions=100/sparseness=1000/other-txn-intents=true/percent-flushed=50-30     670µs ± 3%   620µs ± 4%   -7.44%  (p=0.000 n=9+10)
IntentRangeResolution/versions=100/sparseness=1000/other-txn-intents=true/percent-flushed=100-30   39.1µs ± 1%  11.5µs ± 1%  -70.63%  (p=0.000 n=8+10)
IntentRangeResolution/versions=400/sparseness=1/other-txn-intents=false/percent-flushed=0-30       3.06ms ± 1%  3.07ms ± 2%     ~     (p=0.315 n=10+10)
IntentRangeResolution/versions=400/sparseness=1/other-txn-intents=false/percent-flushed=50-30      4.08ms ± 1%  4.08ms ± 1%     ~     (p=0.842 n=9+10)
IntentRangeResolution/versions=400/sparseness=1/other-txn-intents=false/percent-flushed=100-30     70.8µs ± 2%  78.3µs ± 2%  +10.63%  (p=0.000 n=10+10)
IntentRangeResolution/versions=400/sparseness=100/other-txn-intents=false/percent-flushed=0-30     2.99ms ± 2%  2.99ms ± 3%     ~     (p=0.739 n=10+10)
IntentRangeResolution/versions=400/sparseness=100/other-txn-intents=false/percent-flushed=50-30    3.99ms ± 2%  4.00ms ± 2%     ~     (p=0.436 n=10+10)
IntentRangeResolution/versions=400/sparseness=100/other-txn-intents=false/percent-flushed=100-30   2.20µs ± 3%  2.32µs ± 2%   +5.55%  (p=0.000 n=9+10)
IntentRangeResolution/versions=400/sparseness=100/other-txn-intents=true/percent-flushed=0-30      2.99ms ± 1%  2.95ms ± 2%   -1.57%  (p=0.000 n=10+10)
IntentRangeResolution/versions=400/sparseness=100/other-txn-intents=true/percent-flushed=50-30     4.03ms ± 2%  3.96ms ± 2%   -1.81%  (p=0.004 n=10+9)
IntentRangeResolution/versions=400/sparseness=100/other-txn-intents=true/percent-flushed=100-30    43.2µs ± 2%  12.1µs ± 1%  -71.91%  (p=0.000 n=9+10)
IntentRangeResolution/versions=400/sparseness=1000/other-txn-intents=false/percent-flushed=0-30    2.98ms ± 2%  2.95ms ± 1%     ~     (p=0.052 n=10+10)
IntentRangeResolution/versions=400/sparseness=1000/other-txn-intents=false/percent-flushed=50-30   3.94ms ± 0%  3.93ms ± 1%     ~     (p=0.481 n=8+9)
IntentRangeResolution/versions=400/sparseness=1000/other-txn-intents=false/percent-flushed=100-30  1.37µs ± 2%  1.33µs ± 0%   -2.95%  (p=0.000 n=9+8)
IntentRangeResolution/versions=400/sparseness=1000/other-txn-intents=true/percent-flushed=0-30     3.00ms ± 1%  2.95ms ± 3%   -1.92%  (p=0.001 n=10+10)
IntentRangeResolution/versions=400/sparseness=1000/other-txn-intents=true/percent-flushed=50-30    3.99ms ± 1%  3.94ms ± 2%   -1.24%  (p=0.006 n=9+10)
IntentRangeResolution/versions=400/sparseness=1000/other-txn-intents=true/percent-flushed=100-30   42.8µs ± 3%  11.2µs ± 2%  -73.74%  (p=0.000 n=9+9)
```
